File: modules/todo-storage/android/src/main/java/expo/modules/todostorage/TodoStorageModule.kt
package expo.modules.todostorage

import expo.modules.kotlin.functions.Coroutine
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class TodoStorageModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("TodoStorage")

    AsyncFunction("set") Coroutine { key: String, value: String ->
      val context = appContext.reactContext
        ?: throw IllegalStateException("React context is not available")

      TodoDataStoreManager.setString(context, key, value)
    }

    AsyncFunction("getString") Coroutine { key: String ->
      val context = appContext.reactContext
        ?: throw IllegalStateException("React context is not available")

      TodoDataStoreManager.getString(context, key)
    }

    AsyncFunction("delete") Coroutine { key: String ->
      val context = appContext.reactContext
        ?: throw IllegalStateException("React context is not available")

      TodoDataStoreManager.delete(context, key)
    }
  }
}

File: modules/todo-storage/src/TodoStorageModule.ts
import { NativeModule, requireNativeModule } from "expo";

declare class TodoStorageModule extends NativeModule {
  set(key: string, value: string): Promise<void>;
  getString(key: string): Promise<string | null>;
  delete(key: string): Promise<void>;
}

export default requireNativeModule<TodoStorageModule>("TodoStorage");